<html><head><meta charset='utf-8'><meta name='viewport' content='width=device-width, initial-scale=1'>
<meta name='applicable-device' content='pc'><meta name='keywords' content='电脑,电脑讲解,电脑技术,编程,电脑故障维修java基础(21)异常' />
<script src='../../highlight/highlight.pack.js'></script>
<link rel='stylesheet' type='text/css' href='../../highlight/styles/monokai.css'/>

<link rel='stylesheet' href='../../fenxiang/dist/css/share.min.css'>
<script src='../../fenxiang/src/js/social-share.js'></script>
<script src='../../fenxiang/src/js/qrcode.js'></script>

</head><body><script>hljs.initHighlightingOnLoad();</script><script>
var system ={};  
var p = navigator.platform;       
system.win = p.indexOf('Win') == 0;  
system.mac = p.indexOf('Mac') == 0;  
system.x11 = (p == 'X11') || (p.indexOf('Linux') == 0);     
if(system.win||system.mac||system.xll){
document.write("<link href='../css/3.css' rel='stylesheet' type='text/css'>");}else{ document.write("<link href='../css/3wap.css' rel='stylesheet' type='text/css'>");}</script><script src='../../js/3.js'></script><div class='div2'><div class='heading_nav'><ul><div><li><a href='../../index.html'>首页</a></li>
</div><div onclick='hidden1()' >分享</div>
</ul></div></div>
<div id='heading_nav2'> 
<li class='row' >
<div class='social-share' data-mode='prepend'><a href='javascript:' class='social-share-icon icon-heart'></a></div></li></div><script charset='utf-8' src='../../3/js/hengfu.js'></script><script charset='utf-8' src='../../3/js/hengfu2.js'></script><hr><div class='div1'><div class='biaoti'><center>java基础(21)异常</center></div><div class='banquan'>原文出处:本文由博客园博主爱编程的小灰灰提供。<br/>
原文连接:https://www.cnblogs.com/liuhui0308/p/11654786.html</div><br>
    <h1>1.&nbsp;<strong>异常</strong></h1>
<p><span style="font-family: 宋体;">什么是异常？</span>Java<span style="font-family: 宋体;">代码在运行时期发生的问题就是异常。</span></p>
<p><span style="font-family: 宋体;">在</span>Java<span style="font-family: 宋体;">中，把异常信息封装成了一个类。当出现了问题时，就会创建异常类对象并抛出异常相关的信息（如异常出现的位置、原因等）。</span></p>
<p>&nbsp;</p>
<h2><!--[if !supportLists]-->1.1&nbsp;<!--[endif]--><strong>异常的继承体系</strong></h2>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">在</span>Java<span style="font-family: 宋体;">中使用</span><span style="font-family: 'Times New Roman';">Exception</span><span style="font-family: 宋体;">类来描述异常。</span></p>
<p><strong><img src="./images/java基础(21)异常0.png" alt="" /></strong></p>
<p><span style="font-family: 宋体;">查看</span>API<span style="font-family: 宋体;">中</span><span style="font-family: 'Times New Roman';">Exception</span><span style="font-family: 宋体;">的描述，</span>Exception <span style="font-family: 宋体;">类及其子类是 </span><span style="font-family: 'Times New Roman';">Throwable </span><span style="font-family: 宋体;">的一种形式，它</span>用来<span style="font-family: 宋体;">表示</span>java<span style="font-family: 宋体;">程序中可能会产生的异常</span>，并要求对产生的异常进行合理的异常处理。</p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">继续观察，我们可以发现</span>Exception<span style="font-family: 宋体;">有继承关系，它的父类是</span><span style="font-family: 'Times New Roman';">Throwable</span><span style="font-family: 宋体;">。</span><span style="font-family: 'Times New Roman';">Throwable</span><span style="font-family: 宋体;">是</span>Java <span style="font-family: 宋体;">语言中</span>所有错误或异常的超类，即祖宗类。</p>
<p><strong><img src="./images/java基础(21)异常1.png" alt="" /></strong></p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;<img src="./images/java基础(21)异常2.png" alt="" /></p>
<p><span style="font-family: 宋体;">另外，在异常</span>Exception<span style="font-family: 宋体;">类中，有一个子类要特殊说明一下，</span><span style="font-family: 'Times New Roman';">RuntimeException</span><span style="font-family: 宋体;">子类，</span><span style="font-family: 'Times New Roman';">RuntimeException</span><span style="font-family: 宋体;">及其它的子类只能在</span><span style="font-family: 'Times New Roman';">Java</span><span style="font-family: 宋体;">程序运行过程中出现。</span></p>
<p><strong><img src="./images/java基础(21)异常3.png" alt="" /></strong></p>
<p><span style="font-family: 宋体;">我们再来观察</span>Throwable<span style="font-family: 宋体;">类，能够发现与异常</span><span style="font-family: 'Times New Roman';">Exception</span><span style="font-family: 宋体;">平级的有一个</span><span style="font-family: 'Times New Roman';">Error</span><span style="font-family: 宋体;">，它是</span><span style="font-family: 'Times New Roman';">Throwable</span><span style="font-family: 宋体;">的子类，</span><span style="font-family: 宋体;">它</span>用来<span style="font-family: 宋体;">表示</span>java<span style="font-family: 宋体;">程序中可能会产生的</span>严重<span style="font-family: 宋体;">错误</span><span style="font-family: 宋体;">。解决办法只有一个，修改代码避免</span>Error<span style="font-family: 宋体;">错误的产生。</span></p>
<p><strong><img src="./images/java基础(21)异常4.png" alt="" /></strong></p>
<p><strong>异常继承体系总结：</strong></p>
<blockquote>
<p class="16">Throwable: <span style="font-family: 宋体;">它是所有错误与异常的超类（祖宗类）</span></p>
<p class="16">　　Error <span style="font-family: 宋体;">错误</span></p>
<p class="16">　　Exception <span style="font-family: 宋体;">编译期异常</span><span style="font-family: 'Courier New';">,</span><span style="font-family: 宋体;">进行编译</span><span style="font-family: 'Courier New';">JAVA</span><span style="font-family: 宋体;">程序时出现的问题</span></p>
<p>　　　　RuntimeException <span style="font-family: 宋体;">运行期异常</span><span style="font-family: 'Times New Roman';">, JAVA</span><span style="font-family: 宋体;">程序运行过程中出现的问题</span></p>
</blockquote>
<h2><!--[if !supportLists]-->1.2&nbsp;<!--[endif]--><strong>异常与错误的区别</strong></h2>
<p><strong>异常</strong><span style="font-family: 宋体;">：指程序在编译、运行期间发生了某种异常</span>(XxxException)<span style="font-family: 宋体;">，我们可以对异常进行具体的处理。若不处理异常，程序将会结束运行。</span></p>
<p class="16">异常的产生演示如下：</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args) {
    </span><span style="color: #0000ff;">int</span>[] arr = <span style="color: #0000ff;">new</span> <span style="color: #0000ff;">int</span>[3<span style="color: #000000;">];
    System.out.println(arr[</span>0<span style="color: #000000;">]);
    System.out.println(arr[</span>3<span style="color: #000000;">]);
　　</span><span style="color: #008000;">//</span><span style="color: #008000;"> 该句运行时发生了数组索引越界异常ArrayIndexOutOfBoundsException，由于没有处理异常，导致程序无法继续执行，程序结束。</span>
    System.out.println("over"); <span style="color: #008000;">//</span><span style="color: #008000;"> 由于上面代码发生了异常，此句代码不会执行</span>
}</code></pre>

<p><strong>错误</strong><span style="font-family: 宋体;">：指程序在运行期间发生了某种错误</span>(XxxError)<span style="font-family: 宋体;">，</span><span style="font-family: 'Times New Roman';">Error</span><span style="font-family: 宋体;">错误通常没有具体的处理方式，程序将会结束运行。</span><span style="font-family: 'Times New Roman';">Error</span><span style="font-family: 宋体;">错误的发生往往都是系统级别的问题，都是</span><span style="font-family: 'Times New Roman';">jvm</span><span style="font-family: 宋体;">所在系统发生的，并反馈给</span><span style="font-family: 'Times New Roman';">jvm</span><span style="font-family: 宋体;">的。我们无法针对处理，只能修正代码。</span></p>
<p class="16">错误的产生演示如下：</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args) {
　　</span><span style="color: #0000ff;">int</span>[] arr = <span style="color: #0000ff;">new</span> <span style="color: #0000ff;">int</span>[1024*1024*100<span style="color: #000000;">];
　　</span><span style="color: #008000;">//</span><span style="color: #008000;">该句运行时发生了内存溢出错误OutOfMemoryError，开辟了过大的数组空间，导致JVM在分配数组空间时超出了JVM内存空间，直接发生错误。</span>
}</code></pre>

<h2><!--[if !supportLists]-->1.3&nbsp;<!--[endif]--><strong>异常的产生过程解析</strong></h2>
<p><span style="font-family: 宋体;">先运行下面的程序，程序会产生一个数组索引越界异常</span>ArrayIndexOfBoundsException<span style="font-family: 宋体;">。我们通过图解来解析下异常产生的过程。</span></p>
<p class="16">工具类</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ArrayTools{
    </span><span style="color: #008000;">//</span><span style="color: #008000;">对给定的数组通过给定的角标获取元素。</span>
    <span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">int</span> getElement(<span style="color: #0000ff;">int</span>[] arr,<span style="color: #0000ff;">int</span><span style="color: #000000;"> index)    {
        </span><span style="color: #0000ff;">int</span> element =<span style="color: #000000;"> arr[index];
        </span><span style="color: #0000ff;">return</span><span style="color: #000000;"> element;
    }
}</span></code></pre>

<p class="16">测试类</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ExceptionDemo2 {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args)     {
        </span><span style="color: #0000ff;">int</span>[] arr = {34,12,67<span style="color: #000000;">};
        </span><span style="color: #0000ff;">int</span> num = ArrayTools.getElement(arr,4<span style="color: #000000;">)
        System.out.println(</span>"num="+<span style="color: #000000;">num);
        System.out.println(</span>"over"<span style="color: #000000;">);
    }
}</span></code></pre>

<p class="16">上述程序执行过程图解：</p>
<p><strong><img src="./images/java基础(21)异常5.png" alt="" width="847" height="461" /></strong></p>
<h2><!--[if !supportLists]-->1.4&nbsp;<!--[endif]--><strong><span style="font-family: 华文楷体;">抛出异常</span>throw</strong></h2>
<p>在编写程序时，我们必须要考虑程序出现问题的情况。比如，在定义方法时，方法需要接受参数。那么，当调用方法使用接受到的参数时，首先需要先对参数数据进行合法的判断，数据若不合法，就应该告诉调用者，传递合法的数据进来。这时需要使用抛出异常的方式来告诉调用者。</p>
<p><span style="font-family: 宋体;">在</span>java<span style="font-family: 宋体;">中，提供了一个</span><span style="font-family: 'Times New Roman';">throw</span><span style="font-family: 宋体;">关键字，它用来抛出一个指定的异常对象。那么，抛出一个异常具体如何操作呢？</span></p>
<p class="16">1,创建一个异常对象。封装一些提示信息(信息可以自己编写)。</p>
<p class="16">2,需要将这个异常对象告知给调用者。怎么告知呢？怎么将这个异常对象传递到调用者处呢？通过关键字throw就可以完成。throw 异常对象；</p>
<p>throw<span style="font-family: 宋体;">用在方法内，用来抛出一个异常对象，将这个异常对象传递到调用者处，并结束当前方法的执行。</span></p>
<p class="16">使用格式：</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> 异常类名(参数);</code></pre>

<p>例如：</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> NullPointerException("要访问的arr数组不存在"<span style="color: #000000;">);
</span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> ArrayIndexOutOfBoundsException("该索引在数组中不存在，已超出范围");</code></pre>

<p class="16"><span style="font-family: 宋体;">下面是异常类</span>ArrayIndexOutOfBoundsException与NullPointerException的构造方法</p>
<p><strong><img src="./images/java基础(21)异常6.png" alt="" /></strong></p>
<p class="16"><span style="font-family: 宋体;">学习完抛出异常的格式后，我们通过下面程序演示下</span>throw的使用。</p>
<p class="16">编写工具类，提供获取数组指定索引处的元素值</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ArrayTools{
    </span><span style="color: #008000;">//</span><span style="color: #008000;">通过给定的数组，返回给定的索引对应的元素值。</span>
    <span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">int</span> getElement(<span style="color: #0000ff;">int</span>[] arr,<span style="color: #0000ff;">int</span><span style="color: #000000;"> index)    {
        </span><span style="color: #008000;">/*</span><span style="color: #008000;">
        若程序出了异常，JVM它会打包异常对象并抛出。但是它所提供的信息不够给力。想要更清晰，需要自己抛出异常信息。
下面判断条件如果满足，当执行完throw抛出异常对象后，方法已经无法继续运算。这时就会结束当前方法的执行，并将异常告知给调用者。这时就需要通过异常来解决。
        </span><span style="color: #008000;">*/</span>
        <span style="color: #0000ff;">if</span>(arr==<span style="color: #0000ff;">null</span><span style="color: #000000;">){
            </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> NullPointerException("arr指向的数组不存在"<span style="color: #000000;">);
        }
        </span><span style="color: #0000ff;">if</span>(index&lt;0 || index&gt;=<span style="color: #000000;">arr.length){
            </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> ArrayIndexOutOfBoundsException("错误的角标，"+index+"索引在数组中不存在"<span style="color: #000000;">);
        }
        </span><span style="color: #0000ff;">int</span> element =<span style="color: #000000;"> arr[index];
        </span><span style="color: #0000ff;">return</span><span style="color: #000000;"> element;
    }
}</span></code></pre>

<p class="16">测试类</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ExceptionDemo3 {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args) {
        </span><span style="color: #0000ff;">int</span>[] arr = {34,12,67}; <span style="color: #008000;">//</span><span style="color: #008000;">创建数组</span>
        <span style="color: #0000ff;">int</span> num = ArrayTools.getElement(<span style="color: #0000ff;">null</span>,2);<span style="color: #008000;">//</span><span style="color: #008000;"> 调用方法，获取数组中指定索引处元素
　　　　　</span><span style="color: #008000;">//</span><span style="color: #008000;">int num = ArrayTools.getElement(arr,5);</span><span style="color: #008000;">//</span><span style="color: #008000;"> 调用方法，获取数组中指定索引处元素</span>
        System.out.println("num="+num);<span style="color: #008000;">//</span><span style="color: #008000;">打印获取到的元素值</span>
<span style="color: #000000;">    }
}</span></code></pre>

<h2><!--[if !supportLists]-->1.5&nbsp;<!--[endif]--><strong><span style="font-family: 华文楷体;">声明异常</span>throws</strong></h2>
<p><span style="font-family: 宋体;">声明：将问题标识出来，报告给调用者。如果方法内通过</span>throw<span style="font-family: 宋体;">抛出了编译时异常，而没有捕获处理（稍后讲解该方式），那么必须通过</span><span style="font-family: 'Times New Roman';">throws</span><span style="font-family: 宋体;">进行声明，让调用者去处理。</span></p>
<p>声明异常格式：</p>
<src class="cnblogs_code">
<pre><code>修饰符 返回值类型 方法名(参数) <span style="color: #0000ff;">throws</span> 异常类名1,异常类名2&hellip; {   }</code></pre>

<p>声明异常的代码演示：</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> Demo{
    </span><span style="color: #008000;">/*</span><span style="color: #008000;">
    如果定义功能时有问题发生需要报告给调用者。可以通过在方法上使用throws关键字进行声明。
    </span><span style="color: #008000;">*/</span>
    <span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> show(<span style="color: #0000ff;">int</span> x)<span style="color: #0000ff;">throws</span><span style="color: #000000;"> Exception    {
        </span><span style="color: #0000ff;">if</span>(x&gt;0<span style="color: #000000;">){
            </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span><span style="color: #000000;"> Exception();
        } </span><span style="color: #0000ff;">else</span><span style="color: #000000;"> {
            System.out.println(</span>"show run"<span style="color: #000000;">);
         }
    }
}</span></code></pre>

<p>throws<span style="font-family: 宋体;">用于进行异常类的声明，若该方法可能有多种异常情况产生，那么在</span><span style="font-family: 'Times New Roman';">throws</span><span style="font-family: 宋体;">后面可以写多个异常类，用逗号隔开。</span></p>
<p class="16"><span style="font-family: 宋体;">多个异常的情况，例如</span>:</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">int</span> getElement(<span style="color: #0000ff;">int</span>[] arr,<span style="color: #0000ff;">int</span> index) <span style="color: #0000ff;">throws</span><span style="color: #000000;"> NullPointerException, ArrayIndexOutOfBoundsException {
    </span><span style="color: #0000ff;">if</span>(arr==<span style="color: #0000ff;">null</span><span style="color: #000000;">){
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> NullPointerException("arr指向的数组不存在"<span style="color: #000000;">);
    }
    </span><span style="color: #0000ff;">if</span>(index&lt;0 || index&gt;=<span style="color: #000000;">arr.length){
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> ArrayIndexOutOfBoundsException("错误的角标，"+index+"索引在数组中不存在"<span style="color: #000000;">);
    }
    </span><span style="color: #0000ff;">int</span> element =<span style="color: #000000;"> arr[index];
    </span><span style="color: #0000ff;">return</span><span style="color: #000000;"> element;
}</span></code></pre>

<h2><!--[if !supportLists]-->1.6&nbsp;<!--[endif]--><strong><span style="font-family: 华文楷体;">捕获异常</span>try</strong><strong>&hellip;</strong><strong>catch</strong><strong>&hellip;</strong><strong>finally</strong></h2>
<p><span style="font-family: 宋体;">捕获：</span>Java<span style="font-family: 宋体;">中对异常有针对性的语句进行捕获，可以对出现的异常进行指定方式的处理</span></p>
<p>捕获异常格式：</p>
<blockquote>
<p class="16">c<strong>try：</strong>该代码块中编写可能产生异常的代码。</p>
<p class="16"><strong>catch：</strong>用来进行某种异常的捕获，实现对捕获到的异常进行处理。</p>
<p><strong>finally<span style="font-family: 宋体;">：</span></strong><span style="font-family: 宋体;">有一些特定的代码无论异常是否发生，都需要执行。另外，因为异常会引发程序跳转，导致有些语句执行不到。而</span>finally<span style="font-family: 宋体;">就是解决这个问题的，在</span><span style="font-family: 'Times New Roman';">finally</span><span style="font-family: 宋体;">代码块中存放的代码都是一定会被执行的。</span></p>
</blockquote>
<p><strong>演示如下：</strong></p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ExceptionDemo{
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span> main(String[] args){ <span style="color: #008000;">//</span><span style="color: #008000;">throws ArrayIndexOutOfBoundsException</span>
        <span style="color: #0000ff;">try</span><span style="color: #000000;">    {
              </span><span style="color: #0000ff;">int</span>[] arr = <span style="color: #0000ff;">new</span> <span style="color: #0000ff;">int</span>[3<span style="color: #000000;">];
            System.out.println( arr[</span>5] );<span style="color: #008000;">//</span><span style="color: #008000;"> 会抛出ArrayIndexOutOfBoundsException</span>
<span style="color: #000000;">            当产生异常时，必须有处理方式。要么捕获，要么声明。
        }
        </span><span style="color: #0000ff;">catch</span> (ArrayIndexOutOfBoundsException e) { <span style="color: #008000;">//</span><span style="color: #008000;">括号中需要定义什么呢？try中抛出的是什么异常，在括号中就定义什么异常类型。 </span>
            System.out.println("异常发生了"<span style="color: #000000;">);
        } </span><span style="color: #0000ff;">finally</span><span style="color: #000000;"> {
              arr </span>= <span style="color: #0000ff;">null</span>; <span style="color: #008000;">//</span><span style="color: #008000;">把数组指向null，通过垃圾回收器，进行内存垃圾的清除</span>
<span style="color: #000000;">}
        System.out.println(</span>"程序运行结果"<span style="color: #000000;">);
    }
}</span></code></pre>

<h2>1.7&nbsp;try&hellip;catch&hellip;finally<strong>异常处理的组合方式</strong></h2>
<p class="16"><strong>try catch finally组合：</strong><span style="font-family: 宋体;">检测异常，并传递给</span>catch处理，并在finally中进行资源释放。</p>
<p class="16"><strong>try catch组合</strong>&nbsp;: 对代码进行异常检测，并对检测的异常传递给catch处理。对异常进行捕获处理。</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">void</span> show(){ <span style="color: #008000;">//</span><span style="color: #008000;">不用throws </span>
    <span style="color: #0000ff;">try</span><span style="color: #000000;">{
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> Exception();<span style="color: #008000;">//</span><span style="color: #008000;">产生异常，直接捕获处理</span>
    }<span style="color: #0000ff;">catch</span><span style="color: #000000;">(Exception e){
　　　　</span><span style="color: #008000;">//</span><span style="color: #008000;">处理方式    </span>
<span style="color: #000000;">    }
}</span></code></pre>

<p class="16"><strong><span style="font-family: 宋体;">一个</span>try 多个catch组合</strong>&nbsp;: 对代码进行异常检测，并对检测的异常传递给catch处理。对每种异常信息进行不同的捕获处理。</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">void</span> show(){ <span style="color: #008000;">//</span><span style="color: #008000;">不用throws </span>
    <span style="color: #0000ff;">try</span><span style="color: #000000;">{
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> Exception();<span style="color: #008000;">//</span><span style="color: #008000;">产生异常，直接捕获处理</span>
    }<span style="color: #0000ff;">catch</span><span style="color: #000000;">(XxxException e){
　　　　</span><span style="color: #008000;">//</span><span style="color: #008000;">处理方式    </span>
    }<span style="color: #0000ff;">catch</span><span style="color: #000000;">(YyyException e){
　　　　</span><span style="color: #008000;">//</span><span style="color: #008000;">处理方式    </span>
    }<span style="color: #0000ff;">catch</span><span style="color: #000000;">(ZzzException e){
　　　　</span><span style="color: #008000;">//</span><span style="color: #008000;">处理方式    </span>
<span style="color: #000000;">    }
}</span></code></pre>

<p><span style="font-family: 宋体;">注意</span>:<span style="font-family: 宋体;">这种异常处理方式，要求多个</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">中的异常不能相同，并且若</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">中的多个异常之间有子父类异常的关系，那么子类异常要求在上面的</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">处理，父类异常在下面的</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">处理。</span></p>
<p>&nbsp;</p>
<p class="16"><strong>try finally 组合</strong>: 对代码进行异常检测，检测到异常后因为没有catch，所以一样会被默认jvm抛出。异常是没有捕获处理的。但是功能所开启资源需要进行关闭，所有finally。只为关闭资源。</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">void</span> show(){<span style="color: #008000;">//</span><span style="color: #008000;">需要throws </span>
    <span style="color: #0000ff;">try</span><span style="color: #000000;">{
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span><span style="color: #000000;"> Exception();
    }</span><span style="color: #0000ff;">finally</span><span style="color: #000000;"> {
        </span><span style="color: #008000;">//</span><span style="color: #008000;">释放资源</span>
<span style="color: #000000;">    }
}</span></code></pre>

<p>&nbsp;</p>
<h2><!--[if !supportLists]-->1.8&nbsp;<!--[endif]--><strong>运行时期异常</strong></h2>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p class="16">RuntimeException和他的所有子类异常,都属于运行时期异常。NullPointerException,ArrayIndexOutOfBoundsException等都属于运行时期异常.</p>
<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">运行时期异常的特点</span>:</p>
<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">　　方法中抛出运行时期异常</span>,方法定义中无需throws声明,调用者也无需处理此异常</p>
<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">　　运行时期异常一旦发生</span>,需要程序人员修改源代码.</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ExceptionDemo{
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args){
         method();
    }
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> method(){
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span><span style="color: #000000;"> RuntimeException();
    }
}</span></code></pre>

<p>&nbsp;</p>
<h2><!--[if !supportLists]-->1.9&nbsp;<!--[endif]--><strong>异常在方法重写中细节</strong></h2>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p class="16">子类覆盖父类方法时，如果父类的方法声明异常，子类只能声明父类异常或者该异常的子类，或者不声明。</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">例如：</span></p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method () <span style="color: #0000ff;">throws</span><span style="color: #000000;"> RuntimeException {
　　　}
}
</span><span style="color: #0000ff;">class</span> Zi <span style="color: #0000ff;">extends</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method() <span style="color: #0000ff;">throws</span> RuntimeException { }  <span style="color: #008000;">//</span><span style="color: #008000;">抛出父类一样的异常
    </span><span style="color: #008000;">//</span><span style="color: #008000;">public void method() throws NullPointerException{ } </span><span style="color: #008000;">//</span><span style="color: #008000;">抛出父类子异常</span>
}</code></pre>

<p>&nbsp;</p>
<p class="16">当父类方法声明多个异常时，子类覆盖时只能声明多个异常的子集。</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">例如：</span></p>
<p class="16">&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method () <span style="color: #0000ff;">throws</span><span style="color: #000000;"> NullPointerException, ClassCastException{
　　}
}
</span><span style="color: #0000ff;">class</span> Zi <span style="color: #0000ff;">extends</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method()<span style="color: #0000ff;">throws</span> NullPointerException, ClassCastException { }          <span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method() <span style="color: #0000ff;">throws</span> NullPointerException{ } <span style="color: #008000;">//</span><span style="color: #008000;">抛出父类异常中的一部分</span>
    <span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method() <span style="color: #0000ff;">throws</span> ClassCastException { } <span style="color: #008000;">//</span><span style="color: #008000;">抛出父类异常中的一部分</span>
}</code></pre>

<p class="16">&nbsp;</p>
<p>当被覆盖的方法没有异常声明时，子类覆盖时无法声明异常的。</p>
<p class="16">&nbsp;</p>
<p class="16">&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">例如：</span></p>
<p class="16">&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> method (){
　　}
}
</span><span style="color: #0000ff;">class</span> Zi <span style="color: #0000ff;">extends</span><span style="color: #000000;"> Fu {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method() <span style="color: #0000ff;">throws</span> Exception { }<span style="color: #008000;">//</span><span style="color: #008000;">错误的方式</span>
}</code></pre>

<p class="16">&nbsp;</p>
<p>举例：父类中会存在下列这种情况，接口也有这种情况</p>
<p>问题：接口中没有声明异常，而实现的子类覆盖方法时发生了异常，怎么办？</p>
<p><span style="font-family: 宋体;">答：无法进行</span>throws<span style="font-family: 宋体;">声明，只能</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">的捕获。万一问题处理不了呢？</span><span style="font-family: 'Times New Roman';">catch</span><span style="font-family: 宋体;">中继续</span><span style="font-family: 'Times New Roman';">throw</span><span style="font-family: 宋体;">抛出，但是只能将异常转换成</span><span style="font-family: 'Times New Roman';">RuntimeException</span><span style="font-family: 宋体;">子类抛出。</span></p>
<p class="16">&nbsp;</p>
<p class="16">&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">interface</span><span style="color: #000000;"> Inter {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">abstract</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> method();
}
</span><span style="color: #0000ff;">class</span> Zi <span style="color: #0000ff;">implements</span><span style="color: #000000;"> Inter {
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">void</span> method(){ <span style="color: #008000;">//</span><span style="color: #008000;">无法声明 throws Exception</span>
        <span style="color: #0000ff;">int</span>[] arr = <span style="color: #0000ff;">null</span><span style="color: #000000;">;
        </span><span style="color: #0000ff;">if</span> (arr == <span style="color: #0000ff;">null</span><span style="color: #000000;">) {
            </span><span style="color: #008000;">//</span><span style="color: #008000;">只能捕获处理</span>
            <span style="color: #0000ff;">try</span><span style="color: #000000;">{
</span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> Exception(&ldquo;哥们，你定义的数组arr是空的!<span style="color: #000000;">&rdquo;);
} </span><span style="color: #0000ff;">catch</span><span style="color: #000000;">(Exception e){
    System.out.println(&ldquo;父方法中没有异常抛出，子类中不能抛出Exception异常&rdquo;);
        </span><span style="color: #008000;">//</span><span style="color: #008000;">我们把异常对象e，采用RuntimeException异常方式抛出</span>
        <span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span><span style="color: #000000;"> RuntimeException(e);
　　　　　　}
　　　　}
　　}
}</span></code></pre>

<p>&nbsp;</p>
<h2><!--[if !supportLists]-->1.10&nbsp;<!--[endif]--><strong>异常中常用方法</strong></h2>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">在</span>Throwable<span style="font-family: 宋体;">类中为我们提供了很多操作异常对象的方法，常用的如下：</span></p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;"><img src="./images/java基础(21)异常7.png" alt="" /></span></p>
<p class="16">getMessage方法：返回该异常的详细信息字符串，即异常提示信息</p>
<p class="16">toString方法：返回该异常的名称与详细信息字符串</p>
<p class="16">printStackTrace：在控制台输出该异常的名称与详细信息字符串、异常出现的代码位置</p>
<p class="16">异常的常用方法代码演示：</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">try</span><span style="color: #000000;"> {
    Person p</span>= <span style="color: #0000ff;">null</span><span style="color: #000000;">;
    </span><span style="color: #0000ff;">if</span> (p==<span style="color: #0000ff;">null</span><span style="color: #000000;">) {
        </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span><span style="color: #000000;"> NullPointerException(&ldquo;出现空指针异常了，请检查对象是否为null&rdquo;);
　　}
} </span><span style="color: #0000ff;">catch</span><span style="color: #000000;"> (NullPointerException e) {
    String message </span>=<span style="color: #000000;"> e.getMesage();
    System.out.println(message ); 
    
    String result </span>=<span style="color: #000000;"> e.toString();
    System.out.println(result);    
    
    e.printStackTrace(); 
}</span></code></pre>

<p>&nbsp;</p>
<h1>2.&nbsp;<strong>自定义异常</strong></h1>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">在上述代码中，发现这些异常都是</span>JDK<span style="font-family: 宋体;">内部定义好的，并且这些异常不好找。书写时也很不方便，那么能不能自己定义异常呢？</span></p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">之前的几个异常都是</span>java<span style="font-family: 宋体;">通过类进行的描述。并将问题封装成对象，异常就是将问题封装成了对象。这些异常不好认，书写也很不方便，能不能定义一个符合我的程序要求的异常名称。既然</span><span style="font-family: 'Times New Roman';">JDK</span><span style="font-family: 宋体;">中是使用类在描述异常信息，那么我们也可以模拟</span><span style="font-family: 'Times New Roman';">Java</span><span style="font-family: 宋体;">的这种机制，我们自己定义异常的信息，异常的名字，让异常更符合自己程序的阅读。准确对自己所需要的异常进行类的描述。</span></p>
<p>&nbsp;</p>
<h2><!--[if !supportLists]-->2.1&nbsp;<!--[endif]--><strong>自定义异常类的定义</strong></h2>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">通过阅读异常源代码：发现</span>java<span style="font-family: 宋体;">中所有的异常类，都是继承</span><span style="font-family: 'Times New Roman';">Throwable</span><span style="font-family: 宋体;">，或者继承</span><span style="font-family: 'Times New Roman';">Throwable</span><span style="font-family: 宋体;">的子类。这样该异常才可以被</span><span style="font-family: 'Times New Roman';">throw</span><span style="font-family: 宋体;">抛出。</span></p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">说明这个异常体系具备一个特有的特性：可抛性：即可以被</span>throw<span style="font-family: 宋体;">关键字操作。</span></p>
<p>&nbsp;</p>
<p>并且查阅异常子类源码，发现每个异常中都调用了父类的构造方法，把异常描述信息传递给了父类，让父类帮我们进行异常信息的封装。</p>
<p>&nbsp;</p>
<p>例如NullPointerException<span style="font-family: 宋体;">异常类源代码</span>：</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">class</span> NullPointerException <span style="color: #0000ff;">extends</span><span style="color: #000000;"> RuntimeException {
    </span><span style="color: #0000ff;">public</span><span style="color: #000000;"> NullPointerException() {
        </span><span style="color: #0000ff;">super</span>();<span style="color: #008000;">//</span><span style="color: #008000;">调用父类构造方法</span>
<span style="color: #000000;">    }
    </span><span style="color: #0000ff;">public</span><span style="color: #000000;"> NullPointerException(String s) {
        </span><span style="color: #0000ff;">super</span>(s);<span style="color: #008000;">//</span><span style="color: #008000;">调用父类具有异常信息的构造方法</span>
<span style="color: #000000;">    }
}</span></code></pre>

<p>&nbsp;</p>
<p>现在，我们来定义个自己的异常，即自定义异常。</p>
<p><strong>格式：</strong></p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code>Class 异常名 <span style="color: #0000ff;">extends</span> Exception{ <span style="color: #008000;">//</span><span style="color: #008000;">或继承RuntimeException</span>
    <span style="color: #0000ff;">public</span><span style="color: #000000;"> 异常名(){
　　}
    </span><span style="color: #0000ff;">public</span><span style="color: #000000;"> 异常名(String s){ 
　　　　</span><span style="color: #0000ff;">super</span><span style="color: #000000;">(s); 
　　}
}</span></code></pre>

<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">自定义异常继承</span>Exception演示</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span> MyException <span style="color: #0000ff;">extends</span><span style="color: #000000;"> Exception{
    </span><span style="color: #008000;">/*</span><span style="color: #008000;">
    为什么要定义构造函数，因为看到Java中的异常描述类中有提供对异常对象的初始化方法。
    </span><span style="color: #008000;">*/</span>
    <span style="color: #0000ff;">public</span><span style="color: #000000;"> MyException(){
        </span><span style="color: #0000ff;">super</span><span style="color: #000000;">();
    }
    </span><span style="color: #0000ff;">public</span><span style="color: #000000;"> MyException(String message)    {
        </span><span style="color: #0000ff;">super</span>(message);<span style="color: #008000;">//</span><span style="color: #008000;"> 如果自定义异常需要异常信息，可以通过调用父类的带有字符串参数的构造函数即可。</span>
<span style="color: #000000;">    }
}</span></code></pre>

<p>&nbsp;</p>
<p class="16"><span style="font-family: 宋体;">自定义异常继承</span>RuntimeException演示</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span> MyException <span style="color: #0000ff;">extends</span><span style="color: #000000;"> RuntimeException{
    </span><span style="color: #008000;">/*</span><span style="color: #008000;">
    为什么要定义构造函数，因为看到Java中的异常描述类中有提供对异常对象的初始化方法。
    </span><span style="color: #008000;">*/</span><span style="color: #000000;">
    MyException(){
        </span><span style="color: #0000ff;">super</span><span style="color: #000000;">();
    }
    MyException(String message)    {
        </span><span style="color: #0000ff;">super</span>(message);<span style="color: #008000;">//</span><span style="color: #008000;"> 如果自定义异常需要异常信息，可以通过调用父类的带有字符串参数的构造函数即可。</span>
<span style="color: #000000;">    }
}</span></code></pre>

<p>&nbsp;</p>
<h2><!--[if !supportLists]-->2.2&nbsp;<!--[endif]--><strong>自定义异常的练习</strong></h2>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">定义</span>Person<span style="font-family: 宋体;">类，包含</span><span style="font-family: 'Times New Roman';">name</span><span style="font-family: 宋体;">与</span><span style="font-family: 'Times New Roman';">age</span><span style="font-family: 宋体;">两个成员变量。</span></p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">在</span>Person<span style="font-family: 宋体;">类的有参数构造方法中，进行年龄范围的判断，若年龄为负数或大于</span><span style="font-family: 'Times New Roman';">200</span><span style="font-family: 宋体;">岁，则抛出</span><span style="font-family: 'Times New Roman';">NoAgeException</span><span style="font-family: 宋体;">异常，异常提示信息&ldquo;年龄数值非法&rdquo;。</span></p>
<p>&nbsp;</p>
<p><span style="font-family: 宋体;">要求：在测试类中，调用有参数构造方法，完成</span>Person<span style="font-family: 宋体;">对象创建，并进行异常的处理。</span></p>
<p>&nbsp;</p>
<p class="16">自定义异常类</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span> NoAgeException <span style="color: #0000ff;">extends</span><span style="color: #000000;"> Exception{
    NoAgeException() {
        </span><span style="color: #0000ff;">super</span><span style="color: #000000;">();
    }

    NoAgeException(String message)    {
        </span><span style="color: #0000ff;">super</span><span style="color: #000000;">(message);
    }
}</span></code></pre>

<p>&nbsp;</p>
<p class="16">Person类</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> Person{
    </span><span style="color: #0000ff;">private</span><span style="color: #000000;"> String name;
    </span><span style="color: #0000ff;">private</span> <span style="color: #0000ff;">int</span><span style="color: #000000;"> age;
    Person(String name,</span><span style="color: #0000ff;">int</span> age) <span style="color: #0000ff;">throws</span><span style="color: #000000;"> NoAgeException    {
        </span><span style="color: #008000;">//</span><span style="color: #008000;">加入逻辑判断。</span>
        <span style="color: #0000ff;">if</span>(age&lt;0 || age&gt;200<span style="color: #000000;">)        {
            </span><span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">new</span> NoAgeException(age+",年龄数值非法"<span style="color: #000000;">);
        }
        </span><span style="color: #0000ff;">this</span>.name =<span style="color: #000000;"> name;
        </span><span style="color: #0000ff;">this</span>.age =<span style="color: #000000;"> age;
    }
    </span><span style="color: #008000;">//</span><span style="color: #008000;">定义Person对象对应的字符串表现形式。覆盖Object中的toString方法。</span>
    <span style="color: #0000ff;">public</span><span style="color: #000000;"> String toString()    {
        </span><span style="color: #0000ff;">return</span> "Person[name="+name+",age="+age+"]"<span style="color: #000000;">;
    }
}</span></code></pre>

<p>&nbsp;</p>
<p class="16">测试类</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<src class="cnblogs_code">
<pre><code><span style="color: #0000ff;">class</span><span style="color: #000000;"> ExceptionDemo{
    </span><span style="color: #0000ff;">public</span> <span style="color: #0000ff;">static</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> main(String[] args) {
        </span><span style="color: #0000ff;">try</span><span style="color: #000000;"> {
            Person p </span>= <span style="color: #0000ff;">new</span> Person("xiaoming",20<span style="color: #000000;">);
            System.out.println(p);
        }
        </span><span style="color: #0000ff;">catch</span><span style="color: #000000;"> (NoAgeException ex){
            System.out.println(</span>"年龄异常啦"<span style="color: #000000;">);
        }
        System.out.println(</span>"over"<span style="color: #000000;">);
    }
}</span></code></pre>

<p>&nbsp;</p>
<p><span style="font-family: 宋体;">总结一下，构造函数到底抛出这个</span>NoAgeException<span style="font-family: 宋体;">是继承</span><span style="font-family: 'Times New Roman';">Exception</span><span style="font-family: 宋体;">呢？还是继承</span><span style="font-family: 'Times New Roman';">RuntimeException</span><span style="font-family: 宋体;">呢？</span></p>
<blockquote>
<p class="16"><span style="font-family: 宋体;">继承</span>Exception，必须要throws声明，一声明就告知调用者进行捕获，一旦问题处理了调用者的程序会继续执行。</p>
<p class="16"><span style="font-family: 宋体;">继承</span>RuntimeExcpetion,不需要throws声明的，这时调用是不需要编写捕获代码的，因为调用根本就不知道有问题。一旦发生NoAgeException，调用者程序会停掉，并有jvm将信息显示到屏幕，让调用者看到问题，修正代码。</p>
</blockquote>

</div><hr><script charset='utf-8' src='../../js/sming.js'></script></body></html>